applications/TopologyOptimizationApplication/custom_elements/math_utilities.h

```diff
@@ -0,0 +1,1773 @@
+
+    typedef unsigned int SizeType;
+
+    typedef MathUtils<TDataType> MathUtilsType;
+
+    typedef DenseVector<Vector> Second_Order_Tensor;
+
+    typedef DenseVector<Second_Order_Tensor> Third_Order_Tensor;
+
+    typedef DenseVector<DenseVector<Matrix> > Fourth_Order_Tensor;
+
+    typedef DenseMatrix<Second_Order_Tensor> Matrix_Second_Tensor;
+
+
+    /**
+     * @}
+     */
+    /**
+     * calculates the solutions for a given cubic polynomial equation
+     * 0= a*x^3+b*x^2+c*x+d
+     * @param a coefficient
+     * @param b coefficient
+     * @param c coefficient
+     * @param d coefficient
+     * @param ZeroTol number treated as zero
+     * @return Vector of solutions
+     * WARNING only valid cubic (not quadratic, not linear, not constant) equations with
+     * three real (not complex) solutions
+     */
+
+    static inline bool CardanoFormula(double a, double b, double c, double d, Vector& solution)
+    {
+        solution.resize(3,false);
+        noalias(solution)= ZeroVector(3);
+
+        if(a==0)
+        {
+            std::cout<<"This is not a cubic equation: CardanoFormula"<<std::endl;
+
+            return false;
+        }
+
+        double p= (3.0*a*c-b*b)/(3.0*a*a);
+
+        double q= 2.0*b*b*b/(27.0*a*a*a)-b*c/(3.0*a*a)+d/a;
+
+        double discriminante= p*p*p/27.0+q*q/4.0;
+
+        if(discriminante>0)
+        {
+            return false;
+        }
+
+        if(discriminante==0)
+        {
+            if( a == 0 )
+                return false;
+
+            solution(0)= pow(q/2.0, 1.0/3.0)-b/(3*a);
+            solution(1)= pow(q/2.0, 1.0/3.0)-b/(3*a);
+            solution(2)= pow(-4.0*q, 1.0/3.0)-b/(3*a);
+
+            return true;
+        }
+
+        if(p < 0) return false; //in this case the square roots below will be negative. This substitutes with better efficiency lines 107-110
+
+        solution(0)=
+            -sqrt(-4.0/3.0*p)*cos(1.0/3.0*acos(-q/2.0*sqrt(-27.0/(p*p*p)))+ Globals::Pi / 3.0)
+            -b/(3*a);
+        solution(1)=
+            sqrt(-4.0/3.0*p)*cos(1.0/3.0*acos(-q/2.0*sqrt(-27.0/(p*p*p))))-b/(3*a)
+            ;
+        solution(2)=
+            -sqrt(-4.0/3.0*p)*cos(1.0/3.0*acos(-q/2.0*sqrt(-27.0/(p*p*p)))- Globals::Pi / 3.0)
+            -b/(3*a);
+
+//        if(std::isnan<double>(solution(0)) || std::isnan<double>(solution(1))|| std::isnan<double>(solution(2)))
+//        {
+//            return false;
+//        }
+
+        return true;
+    }
+    /**
+     * @}
+      */
+    /**
+     * calculates Eigenvalues of given square matrix A.
+     * The QR Algorithm with shifts is used
+     * @param A the given square matrix the eigenvalues are to be calculated.
+     * @param tolerance convergence criteria
+     * @param zero number treated as zero
+     * @return Vector of eigenvalues
+     * WARNING only valid for 2*2 and 3*3 Matrices yet
+     */
+
+    static inline Vector EigenValues(const Matrix& A, double tolerance, double zero)
```

Comment:
Idem, use math_utils of the core as much as possible, and if the code does not require template move it to a cpp